gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeBuilder.java

```diff
@@ -72,6 +86,61 @@ public QueryNode build() {
     return executionTree;
   }
 
+  private QueryNode buildExecutionTreeForSameSourceFilterAndSelection(String source) {
+    if (source.equals("QS")) {
+      return buildExecutionTreeForQsFilterAndSelection(source);
+    } else if(source.equals("EDS")) {
+      return buildExecutionTreeForEdsFilterAndSelection(source);
+    } else {
+      throw new UnsupportedOperationException("Unknown Entities data source. No fetcher for this source.");
+    }
+  }
+
+  private QueryNode buildExecutionTreeForQsFilterAndSelection(String source) {
+    int selectionLimit = executionContext.getEntitiesRequest().getLimit();
+    int selectionOffset = executionContext.getEntitiesRequest().getOffset();
+
+    // query-service/Pinot does not support offset when group by is specified. Since we will be
+    // grouping by at least the entity id, we will compute the non zero pagination ourselves. This
+    // means that we need to request for offset + limit rows so that we can paginate appropriately.
+    // Pinot will do the ordering for us.
+    if (selectionOffset > 0) {
+      selectionLimit = selectionOffset + selectionLimit;
+      selectionOffset = 0;
+    }
+
+    QueryNode rootNode = new SelectionAndFilterNode(source, selectionLimit, selectionOffset);
+    if (executionContext.getEntitiesRequest().getOffset() > 0) {
+      rootNode = new PaginateOnlyNode(
+          rootNode,
+          executionContext.getEntitiesRequest().getLimit(),
+          executionContext.getEntitiesRequest().getOffset()
+      );
+    }
+
+    // If the request has an EntityId EQ filter then there's no need for the 2nd request to get the
+    // total entities. So no need to set the TotalFetcherNode
+    if (ExecutionTreeUtils.hasEntityIdEqualsFilter(executionContext)) {
+      executionContext.setTotal(1);
+    } else {
+      rootNode = new TotalFetcherNode(rootNode, source);
+    }
+
+    return rootNode;
+  }
+
+  private QueryNode buildExecutionTreeForEdsFilterAndSelection(String source) {
+    int selectionLimit = executionContext.getEntitiesRequest().getLimit();
+    int selectionOffset = executionContext.getEntitiesRequest().getOffset();
+    List<OrderByExpression> orderBys = executionContext.getEntitiesRequest().getOrderByList();
+
+    QueryNode rootNode = new SelectionAndFilterNode(source, selectionLimit, selectionOffset);
+    // EDS does not seem to do orderby, limiting and offsetting. We will have to do it ourselves.
```

Comment:
Let's log a JIRA for that and fix that next.

Reply:
https://traceableai.atlassian.net/browse/ENG-5715